Specify "streaming" aggregation as the norm. (#623)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

: A set of reports (i.e., measurements) that are aggregated into an aggregate
  result. As defined in {{!VDAF}}.

Batch bucket:

> : Data associated with a given batch allowing the aggregators to perform

Reviewer: * We should try to make it clear here that there is a many-to-one relationship between buckets and batches.

better suited for the asynchronous model. For use cases where datastore
performance is a concern, the synchronous model may be better suited.

In general, reports cannot be aggregated until the Collector specifies a query
and an aggregation parameter. However, depending on the VDAF and batch mode in
use, it is often possible to begin aggregation as soon as reports arrive. For
example, Prio3 has just one valid aggregation parameter (the empty string), and

> thus allows for eager aggregation; and both the time-interval and

Reviewer: Aligns better with discussion on issue.

```suggestion
thus allows for streaming aggregation; and both the time-interval and
```

Author: Hmm, I'm not so sure. I think there are two concepts at play:

* "All-at-once aggregation" vs "streaming aggregation": this refers to whether aggregation happens "all at once", after recovering all of the relevant output shares (as was specified prior to this PR); or if aggregation happens in a "streaming" fashion as output shares are recovered (as is specified in this PR, and implemented in all DAP implementations).
* "Eager aggregation" vs "lazy aggregation": this refers to whether the aggregation process occurs "eagerly" as reports arrive, or if it is performed "lazily" once a collection request arrives. (Lazy aggregation is required if the Collector needs to choose an aggregation parameter.) Which of these is used is dependent on the implementation, both before & after this PR.

Reviewer: You're right, thanks for clarifying.

better suited for the asynchronous model. For use cases where datastore
performance is a concern, the synchronous model may be better suited.

> In general, reports cannot be aggregated until the Collector specifies a query

Reviewer 2: Is this what is meant here?

```suggestion
In general, aggregation cannot complete until the Collector specifies a query
```

`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` so that the
Helper knows it can clean up its state.

#### Batch Buckets {#batch-buckets}

When aggregation recovers an output share, it must be stored into an appropriate
"batch bucket", which is defined in this section. The data stored in a batch
bucket is kept for eventual use in the {{collect-flow}}.

> Batch buckets are indexed by an aggregation parameter, as well as a batch

Reviewer: Since we no longer support multicollection, perhaps it makes sense to remove the aggregation parameter from the index.

Author: Good point!

As written, we do still allow aggregators to aggregate over an arbitrary number of aggregation parameters; the requirement is that we only allow collection of a batch over a single aggregation parameter.

I removed the index & added a few sentences requiring the Leader & Helper to validate that the aggregation parameter in the `CollectionJobReq` & `AggregateShareReq` match what was used for aggregation of the relevant batch.

But this might still be tricky: a naive Helper that permits aggregation of a batch with multiple aggregation parameters, and doesn't account for there being multiple aggregation parameters in its batch bucket implementation, might aggregate into batch buckets "cross-aggregation-parameter." We might want to block this, but I'm not sure this PR is the right place to do so. Or maybe an implementation note is sufficient?

mode-specific "batch bucket identifier":

* For the time-interval batch mode, the batch bucket identifier is an interval
  of time whose beginning is a multiple of the task's `time_precision` and whose
  duration is equal to the task's `time_precision`. The identifier associated
  with a given output share is the unique such interval containing the client

>   timestamp of the report the output share was recovered from.

Reviewer: It might be helpful to give an example:
For example if `time_precision` is 3,600 seconds and the report's `timestamp` is `1729629081`, then the report's batch bucket is `...`

* For the leader-selected batch mode, the batch bucket identifier is a batch ID.
  The identifier associated with a given output share is the batch ID of the

>   related aggregation.

Reviewer: What do you mean by "aggregation"?

A few different pieces of information are associated with each batch bucket:

* An VDAF-specific aggregate share, as defined in {{!VDAF}}.
* A count of a number of reports included in the batch bucket.

> * A 32-byte checksum value.

Reviewer: ```suggestion
* A 32-byte checksum value, as defined below.
```

When processing an output share `out_share`, the following procedure is used to
update the batch buckets:

* Look up the existing batch bucket for the aggregation parameter `agg_param`
  associated with the aggregation job, and the batch bucket identifier
  associated with the aggregation job and output share.
  * If there is no existing batch bucket, initialize a new one. The initial

>     aggregate share value is computed as `Vdaf.agg_init(agg_param)`; the initial

Reviewer: Add reference to section in the VDAF draft that will define this. Draft 13 isn't out yet, but the section number is stable at this point (https://cfrg.github.io/draft-irtf-cfrg-vdaf/draft-irtf-cfrg-vdaf.html#section-5.4).

    count is `0`; and the initial checksum is 32 zero bytes.
* Update the aggregate share `agg_share` to `Vdaf.agg_update(agg_param,
  agg_share, out_share)`.
* Increment the count by 1.
* Update the checksum value to the bitwise-XOR of the checksum value with the
  SHA256 {{!SHS=DOI.10.6028/NIST.FIPS.180-4}} hash of the report ID associated
  with the output share.

> Implementation note: this text describes a single set of values associated with

Reviewer: ```suggestion
Implementation note: this section describes a single set of values associated with
```

each batch bucket. However, implementations are free to shard the aggregate
share/count/checksum values associated with the batch bucket, combining them
back into a single set of values when reading the batch bucket. This may be
useful to avoid write contention. The aggregate share values are combined using

> `Vdaf.merge(agg_param, agg_shares)`, the count values are combined by summing,

Reviewer: Add reference to VDAF spec

"batch bucket", which is defined in this section. The data stored in a batch
bucket is kept for eventual use in the {{collect-flow}}.

Batch buckets are indexed by a batch mode-specific "batch bucket identifier":

* For the time-interval batch mode, the batch bucket identifier is an interval
  of time whose beginning is a multiple of the task's `time_precision` and whose
  duration is equal to the task's `time_precision`. The identifier associated
  with a given output share is the unique such interval containing the client
  timestamp of the report the output share was recovered from; for example, if
  the task's `time_precision` is 1000 seconds and the report's timestamp is
  1729629081, the relevant batch bucket identifier is `(1729629000, 1000)`. (The
  first element of the pair denotes the start of the batch interval and the
  second denotes the duration.)
* For the leader-selected batch mode, the batch bucket identifier is a batch ID.
  The identifier associated with a given output share is the batch ID of the
  related aggregation job.

A few different pieces of information are associated with each batch bucket:

> * An VDAF-specific aggregate share, as defined in {{!VDAF}}.

Reviewer 3: nit: I figure this used to say "an aggregate share ..."
```suggestion
* A VDAF-specific aggregate share, as defined in {{!VDAF}}.
```

### Obtaining Aggregate Shares {#collect-aggregate}

The Leader must compute its own aggregate share, as well as obtaining the
Helper's encrypted aggregate share, before it can complete a collection job.

First, the Leader retrieves all batch buckets ({{batch-buckets}}) associated
with this collection job. The batch buckets to retrieve depend on the batch mode
of this task:

* For time-interval tasks, this is all batch buckets associated with the
  aggregation parameter provided in the `CollectionJobReq`, and whose batch
  bucket identifiers are contained within the batch interval specified in the
  `CollectionJobReq`'s query.
* For leader-selected tasks, this is the batch bucket associated with the
  aggregation parameter provided in the `CollectionJobReq`, and the batch ID the
  Leader has chosen for this collection job.

> The Leader then combines the values inside the batch bucket, as so:

Reviewer: ```suggestion
The Leader then combines the values inside the batch bucket as follows:
```

> * Aggregate shares are combined via `Vdaf.merge(agg_param, agg_shares)`, where

Reviewer: Reference to VDAF draft

requirements for batch parameters following the procedure in
{{batch-validation}}.

Next, the Helper retrieves and combines the batch buckets associated with the

> request using the same process used by the Leader, arriving at its final

Reviewer: Backwards reference
```suggestion
request using the same process used by the Leader ({{XXX}}), arriving at its final
```
